avail-server/src/main/kotlin/com/avail/server/io/files/FileActions.kt

```diff
@@ -0,0 +1,347 @@
+ *   The location in the file to stop overwriting, exclusive. All data from
+ *   this point should be preserved.
+ *
+ * @constructor
+ * Construct a [RemoveRange].
+ *
+ * @param start
+ *   The location in the file to inserting/overwriting the data.
+ * @param end
+ *   The location in the file to stop overwriting, exclusive. All data from
+ *   this point should be preserved.
+ */
+internal class RemoveRange constructor(
+	private val start: Int,
+	private val end: Int) : FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction =
+		file.removeRange(start, end, timestamp)
+
+	override val type: FileActionType = FileActionType.REMOVE_RANGE
+
+	override val isTraced: Boolean = true
+}
+
+/**
+ * `InsertRange` is a [FileAction] that first removes data with from the range
+ * with an exclusive upper bound, then inserts the new data at the position
+ * where the first element was removed.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ *
+ * @property data
+ *   The [ByteArray] that is to be inserted in the file.
+ * @property start
+ *   The location in the file to inserting/overwriting the data.
+ * @property end
+ *   The location in the file to stop overwriting, exclusive. All data from
+ *   this point should be preserved.
+ *
+ * @constructor
+ * Construct an [InsertRange].
+ *
+ * @param data
+ *   The [ByteArray] that is to be inserted in the file.
+ * @param start
+ *   The location in the file to inserting/overwriting the data.
+ * @param end
+ *   The location in the file to stop overwriting, exclusive. All data from
+ *   this point should be preserved.
+ */
+internal class InsertRange constructor(
+	val data: ByteArray,
+	private val start: Int,
+	private val end: Int): FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction =
+		file.insertRange(data, start, end, timestamp)
+
+	override val type: FileActionType = FileActionType.INSERT_RANGE
+
+	override val isTraced: Boolean = true
+}
+
+/**
+ * `NoAction` is a [FileAction] indicates no action should/could be taken.
+ * 
+ * Some `FileAction`s have an inverse action. `NoAction` is used as the inverse
+ * action to `FileAction`s that have no meaningful inverse.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object NoAction: FileAction
+{
+	val tracedAction = TracedAction(0, NoAction, NoAction)
+
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction =
+		tracedAction
+
+	override val type: FileActionType = FileActionType.NO_ACTION
+}
+
+
+/**
+ * `UndoAction` is a [FileAction] that [executes][execute] the
+ * [inverse][TracedAction.reverseAction] for a `FileAction` that is traced as a
+ * [TracedAction].
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object UndoAction: FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.serverFileWrapper.undo()
+		return NoAction.tracedAction
```

Comment:
It is possible to navigate in both directions. This is handled by external stack pointers to avoid polluting the TracedAction stack with non-EditRange FileTypeActions.